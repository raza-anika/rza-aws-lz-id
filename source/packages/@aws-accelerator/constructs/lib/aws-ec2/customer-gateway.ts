/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface CfnCustomerGatewayProps {
  /**
   * Name of the Customer Gateway
   */
  readonly name: string;

  /**
   * Gateway IP address for customer gateway
   */
  readonly ipAddress: string;

  /**
   * Gateway ASN for customer gateway
   */
  readonly bgpAsn: number;

  readonly tags?: cdk.CfnTag[];
}

interface ICustomerGateway extends cdk.IResource {
  /**
   * The identifier of the customer gateway
   *
   * @attribute
   */
  readonly customerGatewayId: string;

  /**
   * The identifier of the customer gateway
   *
   * @attribute
   */
  readonly customerGatewayName: string;
}

/**
 * Class for Customer Gateway
 */
export class CustomerGateway extends cdk.Resource implements ICustomerGateway {
  public readonly customerGatewayId: string;
  public readonly customerGatewayName: string;
  public readonly name: string;

  constructor(scope: Construct, id: string, props: CfnCustomerGatewayProps) {
    super(scope, id);
    this.name = props.name;

    const resource = new cdk.aws_ec2.CfnCustomerGateway(this, 'CustomerGateway', {
      bgpAsn: props.bgpAsn,
      ipAddress: props.ipAddress,
      type: 'ipsec.1',
      tags: props.tags,
    });
    cdk.Tags.of(this).add('Name', this.name);

    this.customerGatewayId = resource.ref;

    this.customerGatewayName = props.name;
  }
}
