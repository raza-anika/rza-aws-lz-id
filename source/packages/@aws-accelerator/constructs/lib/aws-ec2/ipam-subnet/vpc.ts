/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as AWS from 'aws-sdk';
import { IPv4CidrRange } from 'ip-num';

import { throttlingBackOff } from '@aws-accelerator/utils';

interface IVpc {
  /**
   * The CIDRs of the VPC allocated by IPAM
   */
  readonly allocatedCidrs: IPv4CidrRange[];
  /**
   * The friendly name of the VPC
   */
  readonly name: string;
  /**
   * The existing subnets in the VPC
   */
  readonly subnets: Subnet[];
  /**
   * The ID of the VPC
   */
  readonly vpcId: string;
}

interface VpcProps {
  /**
   * The base CIDR pool of the VPC
   */
  readonly basePool: string[];
  /**
   * The ID of the VPC
   */
  readonly vpcId: string;
}

export class Vpc implements IVpc {
  public allocatedCidrs: IPv4CidrRange[] = [];
  public name = '';
  public subnets: Subnet[] = [];
  public readonly vpcId: string;
  private readonly baseRanges: IPv4CidrRange[];
  private readonly ec2 = new AWS.EC2();

  constructor(props: VpcProps) {
    this.vpcId = props.vpcId;
    this.baseRanges = props.basePool.map(item => {
      return IPv4CidrRange.fromCidr(item);
    });
  }

  public async init(): Promise<Vpc> {
    await this.vpcDetails();
    await this.subnetDetails();
    return this;
  }

  private async vpcDetails(): Promise<void> {
    // Get VPC details
    const vpcDetails = await throttlingBackOff(() => this.ec2.describeVpcs({ VpcIds: [this.vpcId] }).promise());

    if (!vpcDetails.Vpcs || !vpcDetails.Vpcs[0].CidrBlockAssociationSet) {
      throw new Error(`Unable to retrieve CIDR block details for VPC ${this.vpcId}`);
    }

    // Determine CIDRs allocated by IPAM
    const allocations: IPv4CidrRange[] = [];
    for (const baseRange of this.baseRanges) {
      for (const item of vpcDetails.Vpcs[0].CidrBlockAssociationSet ?? []) {
        if (!item.CidrBlock) {
          throw new Error(`Unable to retrieve CIDR block for VPC ${this.vpcId}`);
        }

        // Compare VPC CIDR to base CIDR range
        const vpcRange = IPv4CidrRange.fromCidr(item.CidrBlock);
        if (vpcRange.inside(baseRange) || vpcRange.isEquals(baseRange)) {
          allocations.push(vpcRange);
        }
      }
    }
    if (allocations.length === 0) {
      throw new Error(`Unable to determine VPC CIDRs allocated by IPAM for VPC ${this.vpcId}`);
    }

    // Get name tag
    const nameTag = vpcDetails.Vpcs[0].Tags?.filter(item => item.Key === 'Name')[0].Value;
    if (!nameTag) {
      throw new Error(`Unable to retrieve name tag for VPC ${this.vpcId}`);
    }

    // Set property values
    this.allocatedCidrs = allocations;
    this.name = nameTag;
  }

  private validateSubnet(subnet: AWS.EC2.Subnet) {
    if (!subnet.SubnetId) {
      throw new Error(`Unable to retrieve subnet ID`);
    }
    if (!subnet.CidrBlock) {
      throw new Error(`Unable to retrieve CIDR block for subnet ${subnet.SubnetId}`);
    }
  }

  private getSubnets(subnetLists: AWS.EC2.SubnetList | undefined): Subnet[] {
    const subnets: Subnet[] = [];

    for (const subnet of subnetLists ?? []) {
      this.validateSubnet(subnet);

      // Determine if subnet is in scope of IPAM
      const subnetCidr = IPv4CidrRange.fromCidr(subnet.CidrBlock!);
      for (const vpcRange of this.allocatedCidrs) {
        if (subnetCidr.inside(vpcRange) || subnetCidr.isEquals(vpcRange)) {
          // Get name tag
          const nameTag = subnet.Tags?.filter(item => item.Key === 'Name')[0].Value;

          // Push object to array
          subnets.push(
            new Subnet({
              allocatedCidr: subnetCidr,
              mapPublicIpOnLaunch: subnet.MapPublicIpOnLaunch ?? false,
              name: nameTag,
              subnetId: subnet.SubnetId!,
              tags: subnet.Tags,
            }),
          );
        }
      }
    }

    return subnets;
  }

  private async subnetDetails(): Promise<void> {
    let nextToken: string | undefined = undefined;
    const subnets: Subnet[] = [];
    do {
      // Get subnet details
      const page = await throttlingBackOff(() =>
        this.ec2
          .describeSubnets({ Filters: [{ Name: 'vpc-id', Values: [this.vpcId] }], NextToken: nextToken })
          .promise(),
      );

      // Iterate through subnets
      subnets.push(...this.getSubnets(page.Subnets));

      nextToken = page.NextToken;
    } while (nextToken);

    // Set property value
    this.subnets = subnets;
  }
}

interface ISubnet {
  /**
   * The CIDR of the subnet allocated by IPAM
   */
  readonly allocatedCidr: IPv4CidrRange;
  /**
   * The ID of the subnet
   */
  readonly subnetId: string;
  /**
   * The MapPublicIpOnLaunch subnet attribute
   */
  readonly mapPublicIpOnLaunch?: boolean;
  /**
   * The friendly name of the subnet
   */
  readonly name?: string;
  /**
   * Tags associated with the subnet
   */
  readonly tags?: AWS.EC2.TagList;
}

interface SubnetProps {
  /**
   * The CIDR of the subnet allocated by IPAM
   */
  readonly allocatedCidr: IPv4CidrRange;
  /**
   * The ID of the subnet
   */
  readonly subnetId: string;
  /**
   * The MapPublicIpOnLaunch subnet attribute
   */
  readonly mapPublicIpOnLaunch?: boolean;
  /**
   * The friendly name of the subnet
   */
  readonly name?: string;
  /**
   * Tags associated with the subnet
   */
  readonly tags?: AWS.EC2.TagList;
}

export class Subnet implements ISubnet {
  public readonly allocatedCidr: IPv4CidrRange;
  public readonly subnetId: string;
  public readonly mapPublicIpOnLaunch: boolean | undefined;
  public readonly name: string | undefined;
  public readonly tags: AWS.EC2.TagList | undefined;

  constructor(props: SubnetProps) {
    this.allocatedCidr = props.allocatedCidr;
    this.mapPublicIpOnLaunch = props.mapPublicIpOnLaunch;
    this.name = props.name;
    this.subnetId = props.subnetId;
    this.tags = props.tags;
  }
}

export async function vpcInit(props: VpcProps): Promise<Vpc> {
  // Set static properties
  const vpc = new Vpc({ basePool: props.basePool, vpcId: props.vpcId });

  // Set VPC and subnet properties
  await vpc.init();
  return vpc;
}
