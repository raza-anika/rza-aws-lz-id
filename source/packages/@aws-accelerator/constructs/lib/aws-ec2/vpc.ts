/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { IpamAllocationConfig, OutpostsConfig, VirtualPrivateGatewayConfig } from '@aws-accelerator/config';

import { IpamSubnet } from './ipam-subnet';
import { IPrefixList } from './prefix-list';
import { IRouteTable } from './route-table';

export interface ISubnet extends cdk.IResource {
  /**
   * The identifier of the subnet
   *
   * @attribute
   */
  readonly subnetId: string;

  /**
   * The identifier of the subnet
   *
   * @attribute
   */
  readonly subnetArn: string;

  /**
   * The name of the subnet
   *
   * @attribute
   */
  readonly subnetName: string;

  /**
   * The Availability Zone the subnet is located in
   *
   * @attribute
   */
  readonly availabilityZone: string;
}

export interface SubnetProps {
  readonly name: string;
  readonly availabilityZone: string;
  readonly mapPublicIpOnLaunch?: boolean;
  readonly routeTable: IRouteTable;
  readonly vpc: IVpc;
  readonly basePool?: string[];
  readonly ipamAllocation?: IpamAllocationConfig;
  readonly ipv4CidrBlock?: string;
  readonly kmsKey?: cdk.aws_kms.Key;
  readonly logRetentionInDays?: number;
  readonly tags?: cdk.CfnTag[];
  readonly outpost?: OutpostsConfig;
  // readonly nacl: INacl;
}

export class Subnet extends cdk.Resource implements ISubnet {
  public readonly subnetName: string;
  public readonly availabilityZone: string;
  public readonly ipv4CidrBlock: string;
  public readonly mapPublicIpOnLaunch?: boolean;
  public readonly routeTable: IRouteTable;
  public readonly subnetId: string;
  public readonly subnetArn: string;
  public readonly outpostArn?: string;

  constructor(scope: Construct, id: string, props: SubnetProps) {
    super(scope, id);

    this.subnetName = props.name;
    this.availabilityZone = props.availabilityZone;
    this.mapPublicIpOnLaunch = props.mapPublicIpOnLaunch;
    this.routeTable = props.routeTable;
    this.outpostArn = props.outpost?.arn;
    // Determine if IPAM subnet or native
    let resource: cdk.aws_ec2.CfnSubnet | IpamSubnet;

    if (props.ipv4CidrBlock) {
      this.ipv4CidrBlock = props.ipv4CidrBlock;

      resource = new cdk.aws_ec2.CfnSubnet(this, 'Resource', {
        vpcId: props.vpc.vpcId,
        cidrBlock: props.ipv4CidrBlock,
        availabilityZone: props.availabilityZone,
        mapPublicIpOnLaunch: props.mapPublicIpOnLaunch,
        tags: props.tags,
        outpostArn: props.outpost?.arn,
      });

      cdk.Tags.of(this).add('Name', props.name);
      this.subnetId = resource.ref;
      this.subnetArn = cdk.Stack.of(this).formatArn({
        service: 'ec2',
        resource: 'subnet',
        arnFormat: cdk.ArnFormat.SLASH_RESOURCE_NAME,
        resourceName: resource.ref,
      });
    } else {
      if (!props.basePool) {
        throw new Error(
          `Error creating subnet ${props.name}: must specify basePool property when using ipamAllocation`,
        );
      }
      if (!props.ipamAllocation) {
        throw new Error(
          `Error creating subnet ${props.name}: ipamAllocation property must be defined if not specifying ipv4CidrBlock`,
        );
      }
      if (!props.logRetentionInDays) {
        throw new Error(
          `Error creating subnet ${props.name}: logRetentionInDays property must be defined if not specifying ipv4CidrBlock`,
        );
      }
      if (!props.kmsKey) {
        throw new Error(
          `Error creating subnet ${props.name}: kmsKey property must be defined if not specifying ipv4CidrBlock`,
        );
      }

      resource = new IpamSubnet(this, 'Resource', {
        name: props.name,
        availabilityZone: props.availabilityZone,
        basePool: props.basePool,
        ipamAllocation: props.ipamAllocation,
        vpcId: props.vpc.vpcId,
        mapPublicIpOnLaunch: props.mapPublicIpOnLaunch,
        kmsKey: props.kmsKey,
        logRetentionInDays: props.logRetentionInDays,
        tags: props.tags,
        outpostArn: props.outpost?.arn,
      });

      this.ipv4CidrBlock = resource.ipv4CidrBlock;
      this.subnetId = resource.subnetId;
      this.subnetArn = cdk.Stack.of(this).formatArn({
        service: 'ec2',
        resource: 'subnet',
        arnFormat: cdk.ArnFormat.SLASH_RESOURCE_NAME,
        resourceName: resource.subnetId,
      });
    }

    new cdk.aws_ec2.CfnSubnetRouteTableAssociation(this, 'RouteTableAssociation', {
      subnetId: this.subnetId,
      routeTableId: props.routeTable.routeTableId,
    });
  }
}

export interface INatGateway extends cdk.IResource {
  /**
   * The identifier of the NAT Gateway
   *
   * @attribute
   */
  readonly natGatewayId: string;

  /**
   * The name of the NAT Gateway
   *
   * @attribute
   */
  readonly natGatewayName: string;
}

export interface NatGatewayProps {
  readonly name: string;
  readonly subnet: ISubnet;
  readonly tags?: cdk.CfnTag[];
}

export class NatGateway extends cdk.Resource implements INatGateway {
  public readonly natGatewayId: string;
  public readonly natGatewayName: string;

  constructor(scope: Construct, id: string, props: NatGatewayProps) {
    super(scope, id);

    this.natGatewayName = props.name;

    const resource = new cdk.aws_ec2.CfnNatGateway(this, 'Resource', {
      subnetId: props.subnet.subnetId,
      allocationId: new cdk.aws_ec2.CfnEIP(this, 'Eip', {
        domain: 'vpc',
      }).attrAllocationId,
      tags: props.tags,
    });
    cdk.Tags.of(this).add('Name', props.name);

    this.natGatewayId = resource.ref;
  }
}
export interface ISecurityGroup extends cdk.IResource {
  /**
   * ID for the current security group
   * @attribute
   */
  readonly securityGroupId: string;
}

export interface SecurityGroupProps {
  /**
   * The name of the security group. For valid values, see the GroupName
   * parameter of the CreateSecurityGroup action in the Amazon EC2 API
   * Reference.
   *
   * It is not recommended to use an explicit group name.
   *
   * @default If you don't specify a GroupName, AWS CloudFormation generates a
   * unique physical ID and uses that ID for the group name.
   */
  readonly securityGroupName?: string;

  /**
   * A description of the security group.
   *
   * @default The default name will be the construct's CDK path.
   */
  readonly description?: string;

  /**
   * The outbound rules associated with the security group.
   */
  readonly securityGroupEgress?: SecurityGroupEgressRuleProps[];

  /**
   * The inbound rules associated with the security group.
   */
  readonly securityGroupIngress?: SecurityGroupIngressRuleProps[];

  /**
   * The tags that will be attached to the security group
   */
  readonly tags?: cdk.CfnTag[];

  /**
   * The VPC in which to create the security group.
   */
  readonly vpc?: IVpc;

  /**
   * The VPC in which to create the security group.
   */
  readonly vpcId?: string;
}

export interface SecurityGroupIngressRuleProps {
  readonly ipProtocol: string;
  readonly description?: string;
  readonly cidrIp?: string;
  readonly cidrIpv6?: string;
  readonly sourcePrefixList?: IPrefixList;
  readonly sourcePrefixListId?: string;
  readonly sourceSecurityGroup?: ISecurityGroup;
  readonly fromPort?: number;
  readonly toPort?: number;
}

export interface SecurityGroupEgressRuleProps {
  readonly ipProtocol: string;
  readonly description?: string;
  readonly cidrIp?: string;
  readonly cidrIpv6?: string;
  readonly destinationPrefixList?: IPrefixList;
  readonly destinationPrefixListId?: string;
  readonly destinationSecurityGroup?: ISecurityGroup;
  readonly fromPort?: number;
  readonly toPort?: number;
}

export class SecurityGroup extends cdk.Resource implements ISecurityGroup {
  public readonly securityGroupId: string;

  private readonly securityGroup: cdk.aws_ec2.CfnSecurityGroup;

  constructor(scope: Construct, id: string, props: SecurityGroupProps) {
    super(scope, id);

    if (!props.vpc?.vpcId && !props.vpcId) {
      throw new Error(`A property value for vpc or vpcId must be specified`);
    }

    this.securityGroup = new cdk.aws_ec2.CfnSecurityGroup(this, 'Resource', {
      groupDescription: props.description ?? '',
      securityGroupEgress: props.securityGroupEgress,
      securityGroupIngress: props.securityGroupIngress,
      groupName: props.securityGroupName,
      vpcId: props.vpc?.vpcId ?? props.vpcId,
      tags: props.tags,
    });

    if (props.securityGroupName) {
      cdk.Tags.of(this.securityGroup).add('Name', props.securityGroupName);
    }

    this.securityGroupId = this.securityGroup.ref;
  }

  public addIngressRule(id: string, props: SecurityGroupIngressRuleProps) {
    new cdk.aws_ec2.CfnSecurityGroupIngress(this, id, {
      groupId: this.securityGroupId,
      ipProtocol: props.ipProtocol,
      description: props.description,
      cidrIp: props.cidrIp,
      cidrIpv6: props.cidrIpv6,
      sourcePrefixListId: props.sourcePrefixList?.prefixListId,
      sourceSecurityGroupId: props.sourceSecurityGroup?.securityGroupId,
      fromPort: props.fromPort,
      toPort: props.toPort,
    });
  }

  public addEgressRule(id: string, props: SecurityGroupEgressRuleProps) {
    new cdk.aws_ec2.CfnSecurityGroupEgress(this, id, {
      groupId: this.securityGroupId,
      ipProtocol: props.ipProtocol,
      description: props.description,
      cidrIp: props.cidrIp,
      cidrIpv6: props.cidrIpv6,
      destinationPrefixListId: props.destinationPrefixList?.prefixListId,
      destinationSecurityGroupId: props.destinationSecurityGroup?.securityGroupId,
      fromPort: props.fromPort,
      toPort: props.toPort,
    });
  }
}

/**
 * A NetworkAcl
 *
 *
 */
export interface INetworkAcl extends cdk.IResource {
  /**
   * ID for the current Network ACL
   * @attribute
   */
  readonly networkAclId: string;

  /**
   * ID for the current Network ACL
   * @attribute
   */
  readonly networkAclName: string;
}

/**
 * A NetworkAclBase that is not created in this template
 */
abstract class NetworkAclBase extends cdk.Resource implements INetworkAcl {
  public abstract readonly networkAclId: string;
  public abstract readonly networkAclName: string;
}

/**
 * Properties to create NetworkAcl
 */
export interface NetworkAclProps {
  /**
   * The name of the NetworkAcl.
   */
  readonly networkAclName: string;

  /**
   * The VPC in which to create the NetworkACL.
   */
  readonly vpc: IVpc;

  /**
   * The tags which will be attached to the NetworkACL.
   */
  readonly tags?: cdk.CfnTag[];
}

/**
 * Define a new custom network ACL
 *
 * By default, will deny all inbound and outbound traffic unless entries are
 * added explicitly allowing it.
 */
export class NetworkAcl extends NetworkAclBase {
  /**
   * The ID of the NetworkACL
   *
   * @attribute
   */
  public readonly networkAclId: string;

  /**
   * The Name of the NetworkACL
   *
   * @attribute
   */
  public readonly networkAclName: string;

  /**
   * The VPC ID for this NetworkACL
   *
   * @attribute
   */
  public readonly networkAclVpcId: string;

  constructor(scope: Construct, id: string, props: NetworkAclProps) {
    super(scope, id);

    this.networkAclName = props.networkAclName;

    const resource = new cdk.aws_ec2.CfnNetworkAcl(this, 'Resource', {
      vpcId: props.vpc.vpcId,
      tags: props.tags,
    });
    cdk.Tags.of(this).add('Name', props.networkAclName);

    this.networkAclId = resource.ref;
    this.networkAclVpcId = resource.vpcId;
  }

  public associateSubnet(id: string, props: { subnet: ISubnet }) {
    new cdk.aws_ec2.CfnSubnetNetworkAclAssociation(this, id, {
      networkAclId: this.networkAclId,
      subnetId: props.subnet.subnetId,
    });
  }

  public addEntry(
    id: string,
    props: {
      ruleNumber: number;
      protocol: number;
      ruleAction: 'allow' | 'deny';
      egress: boolean;
      cidrBlock?: string;
      icmp?: {
        code?: number;
        type?: number;
      };
      ipv6CidrBlock?: string;
      portRange?: {
        from: number;
        to: number;
      };
    },
  ) {
    new cdk.aws_ec2.CfnNetworkAclEntry(this, id, {
      networkAclId: this.networkAclId,
      ...props,
    });
  }
}

export interface IVpc extends cdk.IResource {
  /**
   * The identifier of the vpc
   *
   * @attribute
   */
  readonly vpcId: string;
}

/**
 * Construction properties for a  VPC object.
 */
export interface VpcProps {
  readonly name: string;
  readonly dhcpOptions?: string;
  readonly enableDnsHostnames?: boolean;
  readonly enableDnsSupport?: boolean;
  readonly instanceTenancy?: 'default' | 'dedicated';
  readonly internetGateway?: boolean;
  readonly ipv4CidrBlock?: string;
  readonly ipv4IpamPoolId?: string;
  readonly ipv4NetmaskLength?: number;
  readonly tags?: cdk.CfnTag[];
  readonly virtualPrivateGateway?: VirtualPrivateGatewayConfig;
}

/**
 * Defines a  VPC object
 */
export class Vpc extends cdk.Resource implements IVpc {
  public readonly name: string;
  public readonly vpcId: string;
  public readonly cidrs: cdk.aws_ec2.CfnVPCCidrBlock[];
  public readonly internetGateway: cdk.aws_ec2.CfnInternetGateway | undefined;
  public readonly internetGatewayAttachment: cdk.aws_ec2.CfnVPCGatewayAttachment | undefined;
  public readonly dhcpOptionsAssociation: cdk.aws_ec2.CfnVPCDHCPOptionsAssociation | undefined;
  public readonly virtualPrivateGateway: cdk.aws_ec2.VpnGateway | undefined;
  public readonly virtualPrivateGatewayAttachment: cdk.aws_ec2.CfnVPCGatewayAttachment | undefined;

  constructor(scope: Construct, id: string, props: VpcProps) {
    super(scope, id);
    this.name = props.name;

    const resource = new cdk.aws_ec2.CfnVPC(this, 'Resource', {
      cidrBlock: props.ipv4CidrBlock,
      enableDnsHostnames: props.enableDnsHostnames,
      enableDnsSupport: props.enableDnsSupport,
      instanceTenancy: props.instanceTenancy,
      ipv4IpamPoolId: props.ipv4IpamPoolId,
      ipv4NetmaskLength: props.ipv4NetmaskLength,
      tags: props.tags,
    });
    cdk.Tags.of(this).add('Name', props.name);

    this.vpcId = resource.ref;
    this.cidrs = [];

    if (props.internetGateway) {
      this.internetGateway = new cdk.aws_ec2.CfnInternetGateway(this, 'InternetGateway', {});

      this.internetGatewayAttachment = new cdk.aws_ec2.CfnVPCGatewayAttachment(this, 'InternetGatewayAttachment', {
        internetGatewayId: this.internetGateway.ref,
        vpcId: this.vpcId,
      });
    }

    if (props.virtualPrivateGateway) {
      this.virtualPrivateGateway = new cdk.aws_ec2.VpnGateway(this, `VirtualPrivateGateway`, {
        amazonSideAsn: props.virtualPrivateGateway.asn,
        type: 'ipsec.1',
      });

      this.virtualPrivateGatewayAttachment = new cdk.aws_ec2.CfnVPCGatewayAttachment(
        this,
        `VirtualPrivateGatewayAttachment`,
        {
          vpnGatewayId: this.virtualPrivateGateway.gatewayId,
          vpcId: this.vpcId,
        },
      );
    }

    if (props.dhcpOptions) {
      this.dhcpOptionsAssociation = new cdk.aws_ec2.CfnVPCDHCPOptionsAssociation(this, 'DhcpOptionsAssociation', {
        dhcpOptionsId: props.dhcpOptions,
        vpcId: this.vpcId,
      });
    }
  }

  public addFlowLogs(options: {
    destinations: ('s3' | 'cloud-watch-logs')[];
    trafficType: 'ALL' | 'REJECT' | 'ACCEPT';
    maxAggregationInterval: number;
    logFormat?: string;
    logRetentionInDays?: number;
    encryptionKey?: cdk.aws_kms.IKey;
    bucketArn?: string;
  }) {
    // Validate maxAggregationInterval
    const maxAggregationInterval = options.maxAggregationInterval;
    if (maxAggregationInterval != 60 && maxAggregationInterval != 600) {
      throw new Error(`Invalid maxAggregationInterval (${maxAggregationInterval}) - must be 60 or 600 seconds`);
    }

    // Destination: CloudWatch Logs
    if (options.destinations.includes('cloud-watch-logs')) {
      if (!options.encryptionKey) {
        throw new Error('encryptionKey not provided for cwl flow log');
      }

      if (!options.logRetentionInDays) {
        throw new Error('logRetentionInDays not provided for cwl flow log');
      }

      const logGroup = new cdk.aws_logs.LogGroup(this, 'FlowLogsGroup', {
        encryptionKey: options.encryptionKey,
        retention: options.logRetentionInDays,
      });

      const role = new cdk.aws_iam.Role(this, 'FlowLogsRole', {
        assumedBy: new cdk.aws_iam.ServicePrincipal('vpc-flow-logs.amazonaws.com'),
      });

      role.addToPrincipalPolicy(
        new cdk.aws_iam.PolicyStatement({
          actions: [
            'logs:CreateLogDelivery',
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:DeleteLogDelivery',
            'logs:DescribeLogGroups',
            'logs:DescribeLogStreams',
            'logs:PutLogEvents',
          ],
          resources: [logGroup.logGroupArn],
        }),
      );

      new cdk.aws_ec2.CfnFlowLog(this, 'CloudWatchFlowLog', {
        deliverLogsPermissionArn: role.roleArn,
        logDestinationType: 'cloud-watch-logs',
        logDestination: logGroup.logGroupArn,
        resourceId: this.vpcId,
        resourceType: 'VPC',
        trafficType: options.trafficType,
        maxAggregationInterval,
        logFormat: options.logFormat,
      });
    }

    // Destination: S3
    if (options.destinations.includes('s3')) {
      new cdk.aws_ec2.CfnFlowLog(this, 'S3FlowLog', {
        logDestinationType: 's3',
        logDestination: `${options.bucketArn}/vpc-flow-logs/`,
        resourceId: this.vpcId,
        resourceType: 'VPC',
        trafficType: options.trafficType,
        maxAggregationInterval,
        logFormat: options.logFormat,
      });
    }
  }

  public addCidr(options: {
    amazonProvidedIpv6CidrBlock?: boolean;
    cidrBlock?: string;
    ipv4IpamPoolId?: string;
    ipv4NetmaskLength?: number;
    ipv6CidrBlock?: string;
    ipv6IpamPoolId?: string;
    ipv6NetmaskLength?: number;
    ipv6Pool?: string;
  }) {
    // This block is required for backwards compatibility
    // with a previous iteration. It appends a number to the
    // logical ID so more than two VPC CIDRs can be defined.
    let logicalId = 'VpcCidrBlock';
    if (this.cidrs.length > 0) {
      logicalId = `VpcCidrBlock${this.cidrs.length}`;
    }

    // Create a secondary VPC CIDR
    this.cidrs.push(
      new cdk.aws_ec2.CfnVPCCidrBlock(this, logicalId, {
        amazonProvidedIpv6CidrBlock: options.amazonProvidedIpv6CidrBlock,
        cidrBlock: options.cidrBlock,
        ipv4IpamPoolId: options.ipv4IpamPoolId,
        ipv4NetmaskLength: options.ipv4NetmaskLength,
        ipv6CidrBlock: options.ipv6CidrBlock,
        ipv6IpamPoolId: options.ipv6IpamPoolId,
        ipv6Pool: options.ipv6Pool,
        vpcId: this.vpcId,
      }),
    );
  }
}
