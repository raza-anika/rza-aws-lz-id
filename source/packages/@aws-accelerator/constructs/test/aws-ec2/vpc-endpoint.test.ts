/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { SecurityGroup } from '../../lib/aws-ec2/vpc';
import { VpcEndpoint, VpcEndpointType } from '../../lib/aws-ec2/vpc-endpoint';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(VpcEndpoint): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

const securityGroup = new SecurityGroup(stack, 'TestSecurityGroup`', {
  securityGroupName: 'TestSecurityGroup',
  description: `AWS Private Endpoint Zone`,
  vpcId: 'Test',
});

new VpcEndpoint(stack, 'VpcEndpoint', {
  vpcId: 'Test',
  vpcEndpointType: VpcEndpointType.GATEWAY,
  service: 'service',
  subnets: ['Test1', 'Test2'],
  securityGroups: [securityGroup],
  privateDnsEnabled: true,
  policyDocument: new cdk.aws_iam.PolicyDocument({
    statements: [
      new cdk.aws_iam.PolicyStatement({
        sid: 'AccessToTrustedPrincipalsAndResources',
        actions: ['*'],
        effect: cdk.aws_iam.Effect.ALLOW,
        resources: ['*'],
        principals: [new cdk.aws_iam.AnyPrincipal()],
        conditions: {
          StringEquals: {
            'aws:PrincipalOrgID': ['organizationId'],
          },
        },
      }),
    ],
  }),
  routeTables: ['Test1', 'Test2'],
});

/**
 * VpcEndpoint construct test
 */
describe('VpcEndpoint', () => {
  snapShotTest(testNamePrefix, stack);
});
