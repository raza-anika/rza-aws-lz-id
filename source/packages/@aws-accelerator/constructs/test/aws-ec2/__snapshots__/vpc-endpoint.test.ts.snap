// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VpcEndpoint Construct(VpcEndpoint):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "TestSecurityGroupDA4B5F83": Object {
      "Properties": Object {
        "GroupDescription": "AWS Private Endpoint Zone",
        "GroupName": "TestSecurityGroup",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestSecurityGroup",
          },
        ],
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "VpcEndpoint80208C18": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "StringEquals": Object {
                  "aws:PrincipalOrgID": Array [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "RouteTableIds": Array [
          "Test1",
          "Test2",
        ],
        "ServiceName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "com.amazonaws.",
              Object {
                "Ref": "AWS::Region",
              },
              ".service",
            ],
          ],
        },
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
  },
}
`;
